python/contracts.py

```diff
@@ -0,0 +1,320 @@
+# coding: utf8
```

Comment:
I believe using DSL to express the attribute contracts can deal with the very complex scene, but users may write a very complex DSL such as `(Int & GreaterEqual(2))| (Float & Greater(3.0) & Less(10.0)) `  and maybe it's very hard to give a clear diagnostic message.

So I issued a simple way as the following prototype:

``` python
class IntList(object):
    def __init__(self):
        self.__name__ = 'Int of list'

    def __instancecheck__(cls, obj):
        return all(isinstance(x, int) for x in obj)

def diag_attribute(name, v, typ):
    if not isinstance(v, typ):
        raise TypeError("argument {0} received unexpected {1}, should be {2}".format(name, str(v), typ.__name__))


diag_attribute("n_classes", 3, int)
diag_attribute("hidden_units", [1,2, "x"], IntList)
```

which output:

``` text
Traceback (most recent call last):
  File "main.py", line 15, in <module>
    diag_attribute("hidden_units", [1,2, "x"], IntList)
  File "main.py", line 11, in diag_attribute
    raise TypeError("argument {0} received unexpected {1}, should be {2}".format(name, str(v), typ.__name__))
TypeError: argument hidden_units received unexpected [1, 2, 'x'], should be IntList
```

and we can return the following diagnostic message to users

```text
argument hidden_units received unexpected [1, 2, 'x'], should be IntList` to user.
```



From this way, we can using Pythoh build-in types to be the expected type, we can also define custom type by override `__instancecheck__`,  for each custom type, we can give one clear diagnostic message like:

```text
argument xxx received unexpected yyy, should be zzz
```

Reply:
I've modified the diagnostic message to be more natural, please see the (modified) original post.